internal/linker/linker.go

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2022, The Garble Authors.
+// See LICENSE for licensing information.
+
+package linker
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"embed"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"github.com/bluekeyes/go-gitdiff/gitdiff"
+	"github.com/gofrs/flock"
+	"io"
+	"io/fs"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+const (
+	MagicValueEnv = "GARBLE_LNK_MAGIC"
+
+	cacheDirName   = ".garble"
+	versionExt     = ".version"
+	garbleCacheDir = "GARBLE_CACHE_DIR"
+)
+
+var (
+	//go:embed patches/*.patch
+	linkerPatchesFs embed.FS
+
+	linkerPatchesVer string
+	linkerPatches    map[string]string
+
+	baseSrcSubdir = filepath.Join("src", "cmd")
+)
+
+func init() {
+	tmpVer, tmpPatch, err := loadPatches(linkerPatchesFs)
+	if err != nil {
+		panic(fmt.Errorf("cannot retrieve patches info: %v", err))
+	}
+	linkerPatchesVer = tmpVer
+	linkerPatches = tmpPatch
+}
+
+func loadPatches(patchesFs fs.FS) (string, map[string]string, error) {
+	versionHash := sha256.New()
+	patches := make(map[string]string)
+	err := fs.WalkDir(patchesFs, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d.IsDir() {
+			return nil
+		}
+		f, err := patchesFs.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		var patchBuf bytes.Buffer
+		if _, err := io.Copy(&patchBuf, f); err != nil {
+			return err
+		}
+
+		patchBytes := patchBuf.Bytes()
+
+		if _, err := versionHash.Write(patchBytes); err != nil {
+			return err
+		}
+
+		files, _, err := gitdiff.Parse(bytes.NewReader(patchBytes))
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if file.IsDelete || file.IsRename {
+				panic("delete and rename patch not supported")
+			}
+
+			patches[file.OldName] = string(patchBytes)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return "", nil, err
+	}
+	return base64.RawStdEncoding.EncodeToString(versionHash.Sum(nil)), patches, nil
+}
+
+func applyPatch(workingDirectory, patch string) error {
```

Comment:
write a TODO about why we have to use `git`, and link to the gitdiff issue that will allow us to use pure Go in the future